zkevm-circuits/src/circuit_tools/constraint_builder.rs

```diff
@@ -1177,184 +1071,62 @@ macro_rules! _require {
         }
     }};
 
-    // Lookup using a tuple
-    ($cb:expr, ($($v:expr),+) => @$tag:expr) => {{
+    // Store tuple
+    ($cb:expr, $values:expr =>> @$tag:expr) => {{
         let description = concat_with_preamble!(
-            "(",
-            $(
-                stringify!($v),
-                ", ",
-            )*
-            ") => @",
+            stringify!($values),
+            " =>> @",
             stringify!($tag),
         );
-        $cb.add_lookup(
+        $cb.store_tuple(
             description,
             $tag,
-            vec![$($v.expr(),)*],
+            $values,
         );
     }};
-    ($cb:expr, $descr:expr, ($($v:expr),+)  => @$tag:expr) => {{
-        $cb.add_lookup(
-            Box::leak($descr.into_boxed_str()),
+    ($cb:expr, $descr:expr, $values:expr =>> @$tag:expr) => {{
+        $cb.store_tuple(
+            Box::leak($descr.to_string().into_boxed_str()),
             $tag,
-            vec![$($v.expr(),)*],
+            $values,
         );
     }};
 
-    // Lookup using an array
-    ($cb:expr, $values:expr => @$tag:expr) => {{
+    // Do lookup
+    ($cb:expr, $values:expr => @$table:expr) => {{
         let description = concat_with_preamble!(
             stringify!($values),
             " => @",
             stringify!($tag),
         );
         $cb.add_lookup(
```

Comment:
Much of the flags are removed, I'm sad but it does become more user friendly based on Yue's feedback. I still suggest preserving a RLC flag that just combine all the `$values` and `$table`, since RLC is a common usage and only one flag would not be too confusing.

Reply:
Lookups are now "raw" lookups, as in the input/table expressions will be used by halo2 pretty much directly. So not sure how rlc encoding with the current `add_lookup` behavior would help much because if you need RLC encoding you can just do it on the inputs of this function (which is what the lookups generated in `add_columns` in the cell manager currently does). I think this also makes sense for users so that `add_lookup` is easy to understand and the values used in the inputs can be specified directly by the user I think still in an easy way (special RLC option vs just doing `rlc::expr` on the values like they are used to with no internal magic).